### DISCUSSION TITLE
Does the Python milvus SDK support asynchronous queries? If supported, how should it be written? Can you provide an example? Thank you very much.

### DISCUSSION THREAD

**Discussion by bestxiaoxiaoming-hm** [milvus-io/milvus#36556] [Q&A and General discussion]
我在使用_async=true的时候，无法从SearchFuture取出结果

**Comment by yhmo** [comment]
MilvusClient doesn't support future results.
The Collection class supports it.  https://github.com/milvus-io/pymilvus/blob/da51ba14edde3c67b682f2d0ae1b85a095815692/pymilvus/orm/collection.py#L816

```
from pymilvus import (
    FieldSchema, CollectionSchema, DataType,
    utility, Collection, connections
)
import random
connections.connect("default", host="localhost", port=19530)

dim = 128
schema = CollectionSchema(fields=[
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=dim),
])

collection_name = "AAA"

utility.drop_collection(collection_name)
collection = Collection(collection_name, schema)
print("collection created")

batch = 1000
collection.insert(data=[
    [i for i in range(batch)],
    [[random.random() for _ in range(dim)] for _ in range(batch)],
])


index_params_ivf={"index_type": "FLAT", "params": {}, "metric_type": "L2"}
collection.create_index("vector", index_params=index_params_ivf)
collection.load()

count = collection.query(expr="", output_fields=["count(*)"], consistency_level = "Strong")
print(count)


future_res = collection.search(data=[[random.random() for _ in range(dim)]],
                            anns_field="vector",
                            limit=10,
                            param={},
                            _async=True,
                            output_fields=["*"])
# do something other
print("=================")

# get the result
results = future_res.result()
for hits in results:
    for hit in hits:
        print(hit)


```

**Comment by lingvisa** [comment]
In Milvus 2.5, does MilvusClient support asynchronous queries? I want to see whether I should use Collection if MilvusClient doesn't support async yet.

**Reply by yhmo** [reply]
In the pymilvus v2.5.3, we provide AsyncMilvusClient for this purpose: https://github.com/milvus-io/pymilvus/blob/v2.5.3/pymilvus/milvus_client/async_milvus_client.py

AsyncMilvusClient uses async/await, you can read the examples to know how to use it:
https://github.com/milvus-io/pymilvus/blob/v2.5.3/examples/simple_async.py
https://github.com/milvus-io/pymilvus/blob/v2.5.3/examples/simple_async2.py

**Reply by lingvisa** [reply]
That's very cool! By default, AsyncMilvusClient should be used. right?

**Reply by yhmo** [reply]
Both optional.
If you want sync call, use the MilvusClient:
```
milvus_client = MilvusClient(uri=uri)
```
If you want async call, use AsyncMilvusClient
```
milvus_client = AsyncMilvusClient(uri=uri)
```

**Comment by michaelgrewal** [comment]
Can we get support for an async search_iterator() for the AsyncMilvusClient?

**Reply by yhmo** [reply]
So far as I know, search_iteartor/query_iterator are not implemented for AsyncMilvusClient.

**Reply by michaelgrewal** [reply]
It's not implemented, but I'm requesting this feature to be added please

**Reply by yhmo** [reply]
You can add a feature request in this page: https://github.com/milvus-io/pymilvus/issues
To support search_iterator for AsyncMilvusClient, the AsyncMilvusClient.search_iterator() should return an AsyncSearchIterator object, and the AsyncSearchIterator.next() method should return an AsyncSearchPage object.

**Reply by michaelgrewal** [reply]
Thank you, I'll open an issue soon!

**Reply by xiaofan-luan** [reply]
Don't think iterator can be done in a async style.

iterator has states, request need to be issued with the last request done.

**Reply by xiaofan-luan** [reply]
also, I don't see a use case where we need async iterator.

usually there is consumer iterate and consume the data, all should be in sync mode. I would recommended you to have a thread for that if necessary

**Reply by michaelgrewal** [reply]
> iterator has states, request need to be issued with the last request done.

That is OK, but the generator should async and non blocking. Currently, the MilvusClient.search_iterator() returns a synchronous blocking iterator.

My use case and goal is to be able to call on each iterator iteration with async, so that we can run some code concurrently while moving to next iterations.

For example, 
```
tasks = []
async for batch in generator:
    tasks.append(asyncio.create_task(chain.arun(inputs=batch))
results = await asyncio.gather(*tasks)
```
